deque.go

```diff
@@ -432,3 +435,50 @@ func (q *Deque[T]) resize(newSize int) {
 	q.tail = q.count
 	q.buf = newBuf
 }
+
+// TakeSlice returns the contents of the queue as a contiguous slice,
+// leaving the queue empty.
+func (q *Deque[T]) TakeSlice() []T {
+	if q == nil {
+		return nil
+	}
+	if q.head == 0 {
+		// buffer is in correct order, no need to move memory
+	} else if q.head < q.tail {
+		// ..ABCDEFGH..
+		copy(q.buf, q.buf[q.head:q.tail])
+		// ABCDEFGHgh..
+		clear(q.buf[q.count:q.tail])
+		// ABCDEFGH....
```

Comment:
I don't think the copy and clear are needed. Just return the range within `buf`:
```go
buf = q.buf[head:tail]
```

Reply:
Didn't do it this way, because this would make `[0:head]` a wasted memory chunk that can't be deallocated until the buffer's content is moved away via `append` resize or GC collecting the buffer.

If that's not a concern then there are more optimisations possible that waste some `[0:n]` memory in exchange for less work.